Preliminary cleanup of `WitnessPat` hoisting/printing

Zalathar opened the pull request:
Follow-up to #128430.

The eventual goal is to remove `print::Pat` entirely, but in the course of working towards that I made so many small improvements that it seems wise to let those be reviewed/merged on their own first.

Best reviewed commit-by-commit, most of which should be pretty simple and straightforward.

r? @Nadrieril

Zalathar committed:
Split out hoisting/printing of `box` patterns

Zalathar committed:
Split out a `hoist` helper in `hoist_witness_pat`

Zalathar committed:
Break up `print::Pat` printing into several helper functions

Zalathar committed:
Unify `Variant` and `Leaf` into `print::PatKind::StructLike`

Zalathar committed:
Replace an unnecessary slice pattern with `has_dot_dot: bool`

Zalathar committed:
Remove an impossible case under `EnumInfo::NotEnum`

Zalathar committed:
Simplify hoisting of struct-like patterns

Zalathar committed:
Simplify hoisting of array/slice patterns

We can replace some tricky iterator-mutation code with a much simpler version
that uses `while let` to shrink a slice.

We also check whether a subpattern would be a wildcard _before_ hoisting it,
which will be very useful when trying to get rid of `print::PatKind` later.

Zalathar committed:
Simplify hoisting of ref patterns (`&` and `&mut`)

Zalathar committed:
Avoid using `ty::tls::with` in `write_struct_like`

Zalathar committed:
Use `TyCtxt::is_diagnostic_item`

Nadrieril commented on compiler/rustc_pattern_analysis/src/rustc.rs:
```diff
@@ -862,37 +861,35 @@ impl<'p, 'tcx: 'p> RustcPatCtxt<'p, 'tcx> {
+
+                let (mut prefix, mut suffix) = pat.fields.split_at(prefix_len);
+
+                // If the pattern contains a `..`, but is also an array pattern with a fixed
+                // length, then we can slightly simplify diagnostics by merging any adjacent
+                // wildcard patterns into the `..`: `[x, _, .., _, y]` => `[x, .., y]`.
+                // (This simplification isn't allowed for variable-length patterns, because
+                // it would make them accept shorter lists and thus change their meaning.)
```
This is confusing to me, because "array pattern" and "variable-length pattern" sound like they refer to "patterns like `[a, b, c]`" and "patterns like `[a, .., b, c]`" respectively, which isn't what you intended.

```suggestion
                // If the pattern contains a `..` and applies to values of statically-known length,
                // then we can slightly simplify diagnostics by merging any adjacent
                // wildcard patterns into the `..`: `[x, _, .., _, y]` => `[x, .., y]`.
                // (This simplification isn't allowed for slices, because then
                // `[x, .., y]` matches more values than `[x, _, .., _, y]`)
```

Nadrieril commented on compiler/rustc_pattern_analysis/src/rustc.rs:
```diff
@@ -910,6 +907,22 @@ impl<'p, 'tcx: 'p> RustcPatCtxt<'p, 'tcx> {
+            lo: MaybeInfiniteInt::NegInfinity,
+            hi: MaybeInfiniteInt::PosInfinity,
+        })
+        | Constructor::Never
```
`!` should not count as a wildcard here, since `[!, ..]` is semantically different from `[..]`.

```suggestion
```

Nadrieril replied:
If you really want precision, you can do
```suggestion
        | Constructor::Never if tcx.features().never_patterns
```

Zalathar replied:
I'll stick with the less-precise version for now, because requiring `tcx` is going to be somewhat awkward (temporarily) in an upcoming change.

Is it possible for `Constructor::Never` patterns to exist when the feature is disabled? It wasn't obvious to me why the print behaviour is conditional.

Zalathar replied:
Never mind; turns out my future changes will need the precise version to avoid breaking existing tests, and supplying `tcx` is actually fine.

Nadrieril reviewed (changes requested):
This was very pleasant to review, tyvm!

Zalathar commented:
(Rebased before addressing feedback; no changes yet.)

Zalathar commented:
Addressed feedback [(diff)](https://github.com/rust-lang/rust/compare/49306def5a200922335399210f3899558675f809..482412c98ad714ae386a85fc44222197be3d3ffc).

Nadrieril commented:
Ty!

@bors r+

bors commented:
:pushpin: Commit 482412c98ad714ae386a85fc44222197be3d3ffc has been approved by `Nadrieril`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Nadrieril 482412c98ad714ae386a85fc44222197be3d3ffc -->
<!-- homu: {"type":"Approved","sha":"482412c98ad714ae386a85fc44222197be3d3ffc","approver":"Nadrieril","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
